File: wrappers/java/src/test/java/org/hyperledger/indy/sdk/signus/ReplaceKeysTest.java
package org.hyperledger.indy.sdk.signus;

import org.bitcoinj.core.Base58;
import org.hyperledger.indy.sdk.ErrorCode;
import org.hyperledger.indy.sdk.ErrorCodeMatcher;
import org.hyperledger.indy.sdk.IndyIntegrationTest;
import org.hyperledger.indy.sdk.signus.SignusResults.CreateAndStoreMyDidResult;
import org.hyperledger.indy.sdk.wallet.Wallet;

import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.util.concurrent.ExecutionException;

public class ReplaceKeysTest extends IndyIntegrationTest {

	private Wallet wallet;
	private String did;
	private String verkey;
	private String walletName = "signusWallet";

	@Before
	public void createWalletWithDid() throws Exception {
		Wallet.createWallet("default", walletName, "default", null, null).get();
		wallet = Wallet.openWallet(walletName, null, null).get();

		CreateAndStoreMyDidResult result = Signus.createAndStoreMyDid(this.wallet, "{}").get();

		did = result.getDid();
		verkey = result.getVerkey();
	}

	@After
	public void deleteWallet() throws Exception {
		wallet.closeWallet().get();
		Wallet.deleteWallet(walletName, null).get();
	}

	@Test
	public void testReplaceKeysWorksForEmptyJson() throws Exception {
		SignusResults.ReplaceKeysResult result = Signus.replaceKeys(wallet, did, "{}").get();
		assertNotNull(result);

		assertEquals(32, Base58.decode(result.getVerkey()).length);
	}

	@Test
	public void testReplaceKeysWorksForInvalidDid() throws Exception {
		thrown.expect(ExecutionException.class);
		thrown.expectCause(new ErrorCodeMatcher(ErrorCode.CommonInvalidStructure));

		Signus.replaceKeys(this.wallet, "invalid_base58_string", "{}").get();
	}

	@Test
	public void testReplaceKeysWorksForNotExistsDid() throws Exception {
		SignusResults.ReplaceKeysResult result = Signus.replaceKeys(this.wallet, "8wZcEriaNLNKtteJvx7f8i", "{}").get();
		assertNotNull(result);
	}

	@Test
	public void testReplaceKeysWorksForSeed() throws Exception {
		SignusResults.ReplaceKeysResult result = Signus.replaceKeys(this.wallet, this.did, "{\"seed\":\"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa\"}").get();
		assertNotNull(result);
		String verkey = result.getVerkey();

		assertEquals("CnEDk9HrMnmiHXEV1WFgbVCRteYnPqsJwrTdcZaNhFVW", verkey);
		assertNotEquals(this.verkey, verkey);
	}
}

File: wrappers/dotnet/indy-sdk-dotnet-test/SignusTests/ReplaceKeysTest.cs
using Hyperledger.Indy.Test.Util.Base58Check;
using Hyperledger.Indy.SignusApi;
using Hyperledger.Indy.WalletApi;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Threading.Tasks;

namespace Hyperledger.Indy.Test.SignusTests
{
    [TestClass]
    public class ReplaceKeysTest : IndyIntegrationTestBase
    {
        private Wallet _wallet;
        private string _walletName = "SignusWallet";
        private string _did;
        private string _verKey;

        [TestInitialize]
        public async Task CreateWalletWithDid()
        {
            await Wallet.CreateWalletAsync("default", _walletName, "default", null, null);
            _wallet = await Wallet.OpenWalletAsync(_walletName, null, null);

            var result = await Signus.CreateAndStoreMyDidAsync(_wallet, "{}");

            _did = result.Did;
            _verKey = result.VerKey;
        }

        [TestCleanup]
        public async Task DeleteWallet()
        {
            if(_wallet != null)
                await _wallet.CloseAsync();

            await Wallet.DeleteWalletAsync(_walletName, null);
        }
        
        [TestMethod]
        public async Task TestReplaceKeysWorksForEmptyJson()
        {
            var result = await Signus.ReplaceKeysAsync(_wallet, _did, "{}");

            Assert.IsNotNull(result);
            Assert.AreEqual(32, Base58CheckEncoding.DecodePlain(result.VerKey).Length);
        }

        [TestMethod]
        public async Task TestReplaceKeysWorksForInvalidDid()
        {
            var ex = await Assert.ThrowsExceptionAsync<IndyException>(() =>
                Signus.ReplaceKeysAsync(_wallet, "invalid_base58_string", "{}")
            );

            Assert.AreEqual(ErrorCode.CommonInvalidStructure, ex.ErrorCode);
        }

        [TestMethod]
        public async Task TestReplaceKeysWorksForNotExistsDid()
        {
            var result = await Signus.ReplaceKeysAsync(_wallet, "8wZcEriaNLNKtteJvx7f8i", "{}");

            Assert.IsNotNull(result);
        }

        [TestMethod]
        public async Task TestReplaceKeysWorksForSeed()
        {
            var result = await Signus.ReplaceKeysAsync(_wallet, _did, "{\"seed\":\"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa\"}");

            Assert.IsNotNull(result);
            Assert.AreEqual("CnEDk9HrMnmiHXEV1WFgbVCRteYnPqsJwrTdcZaNhFVW", result.VerKey);
            Assert.AreNotEqual(_verKey, result.VerKey);
        }

    }
}
